src/org/intellij/grammar/actions/BnfGenerateLexerAction.java

```diff
@@ -172,24 +179,33 @@ public void visitElement(@NotNull PsiElement element) {
     }
     catch (Throwable ignore) {}
     ve.init();
-    
+
     VelocityContext context = new VelocityContext();
     context.put("lexerClass", getLexerName(bnfFile));
-    context.put("packageName", StringUtil.notNullize(packageName, StringUtil.getPackageName(getRootAttribute(bnfFile, KnownAttribute.PARSER_CLASS))));
+    putPackageName(context, bnfFile, packageName);
     context.put("tokenPrefix", getRootAttribute(bnfFile, KnownAttribute.ELEMENT_TYPE_PREFIX));
-    context.put("typesClass", getRootAttribute(bnfFile, KnownAttribute.ELEMENT_TYPE_HOLDER_CLASS));
+    putTypeHolderClass(context, bnfFile);
     context.put("tokenPrefix", getRootAttribute(bnfFile, KnownAttribute.ELEMENT_TYPE_PREFIX));
     context.put("simpleTokens", simpleTokens);
     context.put("regexpTokens", regexpTokens);
     context.put("StringUtil", StringUtil.class);
     context.put("maxTokenLength", maxLen[0]);
 
     StringWriter out = new StringWriter();
-    InputStream stream = getClass().getResourceAsStream("/templates/lexer.flex.template");
+    InputStream stream = getClass().getResourceAsStream(getLexerFlexTemplate());
     ve.evaluate(context, out, "lexer.flex.template", new InputStreamReader(stream));
     return StringUtil.convertLineSeparators(out.toString());
   }
 
+  protected void putPackageName(@NotNull VelocityContext context, BnfFile bnfFile, @Nullable String packageName) {
```

Comment:
Adding a method for each property is not good. It's a famous OO-spaghetti code.


One method for context creation would be better.